### DISCUSSION TITLE
Pre/post processing

### DISCUSSION THREAD

**Discussion by codebutler** [rpldy/react-uploady#492] [Q&A]
Hi,

How can I run async functions before and after each item upload? The behavior I have in mind is:

* Item stays in "uploading" (or some sort of loading) state from just before preprocess function runs until after postprocess function finishes.
    * Currently PRE_SEND seems to happen in the "added" state and ITEM_FINALIZE in the "finished" state
* Any errors thrown should put the item into "error" state.
    * Currently errors thrown in a PRE_SEND callback do put the item into "error" state, but errors thrown in the ITEM_FINALIZE callback do not.
* Some way to pass data from the preprocess function to the postprocess function. 

Thanks

**Comment by yoavniran** [comment]
hi @codebutler 
I'd like to clarify, it sounds like you're asking for a way to throw an error in the FINALIZE event handler in order to change the state of the item. Is that correct?
If so, it is not possible, by design. FINALIZE happens after Uploady finished handling an item (successfully or not) and doesnt make (or allow) any more state changes. 

What's the use-case here? Why would you want to change a successful item into a failed one? And why would you need to do it with Uploady and not in your own state if you have some post-processing taking place after the upload finished. Remember that as far as Uploady is concerned, if the server reported the file uploaded successfully, then the item's state is FINISHED (with success).
